### ISSUE TITLE
destfile property is not honored when a Premain Agent class is instrumented

### ISSUE THREAD

**Issue by rshivane** [jacoco/jacoco#597]
### Steps to reproduce

JaCoCo version: 0.7.9
Operating system: Windows
Tool integration: Maven

##### Background
My project generates a jar with a Java Premain Agent ([refer pom.xml](https://github.com/ApptuitAI/JInsight/blob/de36b148443b3dd267f4b2cf17178212122803e9/pom.xml#L307))

The [pom.xml is setup to do offline instrumentation](https://github.com/ApptuitAI/JInsight/blob/de36b148443b3dd267f4b2cf17178212122803e9/pom.xml#L396) of the project.

The surefire plugin is configured to [set the destfile of jacoco.exec to the target folder](https://github.com/ApptuitAI/JInsight/blob/de36b148443b3dd267f4b2cf17178212122803e9/pom.xml#L294):
```xml
        <configuration>
          <systemPropertyVariables>
            <jacoco-agent.destfile>${project.build.directory}/jacoco.exec</jacoco-agent.destfile>
          </systemPropertyVariables>
```

##### Issue
However, when I run the tests, the exec file is always written to the working directory and not the target directory.

##### Analysis
Attaching the process indicates that the System property is not available in the premain method of the agent. Also the AgentOptions has size zero. Screenshot:
![image](https://user-images.githubusercontent.com/31876561/30743670-3d483abe-9fbc-11e7-9292-cdfa6ad69825.png)

I wonder if jacoco runtime is being initialized only *after* premain is complete.

##### Expected
For now, I have [excluded the my Agent class](https://github.com/ApptuitAI/JInsight/blob/de36b148443b3dd267f4b2cf17178212122803e9/pom.xml#L402) from offline instrumentation. 

This gives me no visibility into the code coverage. I would *love* to re-enable offline instrumentation for the class.

**Comment by rshivane** [comment]
Interestingly, moving the destfile property from ```<systemPropertyVariables><jacoco-agent.destfile>``` to ``` <argLine> ``` fixed the issue

![image](https://user-images.githubusercontent.com/31876561/30744592-5d53d008-9fc0-11e7-8b96-8602d25f8dfb.png)

Complete list of changes to pom.xml to make this work: https://github.com/ApptuitAI/JInsight/commit/b4f9f9cbc9523d8f5acf2c0ee501a29e3d020cc4

**Comment by Godin** [comment]
Similarly to #591 :

`maven-surefire-plugin` documentation last paragraph on page http://maven.apache.org/surefire/maven-surefire-plugin/examples/system-properties.html states:

> Special VM Properties
Some system properties must be set on the command line of the forked VM, and cannot be set after the VM has been started. These properties must be added to the argLine parameter of the Surefire plugin.

Rephrasing: system properties specified via `<systemProperties>` of `maven-surefire-plugin` are set after JVM startup, because not passed on command line - see https://github.com/apache/maven-surefire/blob/surefire-2.20/surefire-booter/src/main/java/org/apache/maven/surefire/booter/PropertiesWrapper.java#L50-L56

Startup of any Java agent happens after JVM startup and before execution of main, i.e. "surefire booter" is executed after startup of Java agent and so properties specified via `systemProperties` won't be visible in agent.

To sum up - solutions:
* specify `jacoco-agent.destfile` using `argLine`
* or via `jacoco-agent.properties` (see http://www.jacoco.org/jacoco/trunk/doc/offline.html)
* or why not use on-the-fly instrumentation with JaCoCo agent instead of offline, agents are applied in the order they are specified on command-line, so JaCoCo agent simply should be the first one

**Comment by rshivane** [comment]
Thanks for the details Godin. 

> why not use on-the-fly instrumentation with JaCoCo agent

When I setup on-the-fly instrumentation, the biggest roadblock for me is, the running/debugging maven-surefire tests from IntelliJ IDEA fails. I get the error `Error: Could not find or load main class @{argLine}` when I run the test from IDEA

**Comment by Godin** [comment]
@rshivane This is interesting - wasn't aware that nowadays IntelliJ IDEA respects configuration of `maven-surefire-plugin`. And what about configuration without use of `@{argLine}` as in http://www.jacoco.org/jacoco/trunk/doc/prepare-agent-mojo.html :
```
  <properties>
    <argLine>-your -extra -arguments</argLine>
  </properties>
  ...
  <plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-surefire-plugin</artifactId>
    <configuration>
      <!-- no argLine here -->
    </configuration>
  </plugin>
```
? Or a separate profile that won't be active in IDE?
